### PR TITLE
Documentation: add 'why.md'

### PR DESCRIPTION
**Pull request by gyuho**

### CODE REVIEW

File: Documentation/learning/why.md [hunk 1]
```diff
@@ -0,0 +1,6 @@
+# When to use etcd
```
Comment (xiang90) [inline Documentation/learning/why.md]: # Why etcd

**Comment by gyuho** [comment]
All done. Thanks!

**Comment by xiang90** [comment]
defer to @fanminshi and @heyitsanthony

File: Documentation/learning/why.md [hunk 2]
```diff
@@ -0,0 +1,6 @@
+# Why etcd
+
+etcd is designed to store a small amount of metadata in consistent, fault-tolerant way. Distributed systems use etcd as the root of scheduling and service discovery configuration storage. [Distributed lock][etcd-concurrency] is implemented on top of etcd, useful for master election in other distributed databases.
```
Comment (heyitsanthony) [inline Documentation/learning/why.md]: what's meant to be difference between this and https://github.com/coreos/etcd/blob/master/README.md#etcd ?

Comment (gyuho) [inline Documentation/learning/why.md]: Yeah should we just drop it and move @fanminshi 's `origin of etcd` to main README? 
I don't have strong preference.

/cc @xiang90

Comment (xiang90) [inline Documentation/learning/why.md]: This doc eventually will be more extensive. We will have the feature charts, comparison, use case shows, etc..

The readme one should be a one sentence explanation.

Comment (xiang90) [inline Documentation/learning/why.md]: @gyuho I do not want to make readme super long. actually now, it is already longer than I would expect.

Comment (fanminshi) [inline Documentation/learning/why.md]: I agree with xiang90. i think readme should be a short introduction about etcd. and why etcd should go in depth on what's etcd, its benefits over other solutions, and etc.

Comment (heyitsanthony) [inline Documentation/learning/why.md]: can those be marked as sections with TODO's, then? it's not clear why this exists without that

File: Documentation/learning/why.md [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+
+# What does etcd stand for?
+
+TODO(fanmin)
+
+# Who uses etcd: container linux by CoreOS
```
Comment (xiang90) [inline Documentation/learning/why.md]: Example use cases:

- container linux by CoreOS

File: Documentation/learning/why.md [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+
+Application running on [container linux][container-linux] gets automatic, no-downtime Linux kernel updates. CoreOS uses [locksmith] to coordinate updates. locksmith stores semaphore values in etcd to ensure that only subset of cluster are rebooting at any given time.
+
```
Comment (xiang90) [inline Documentation/learning/why.md]: - Kubernetes

File: Documentation/learning/why.md [hunk 5]
```diff
@@ -0,0 +1,21 @@
+# Why etcd
+
+etcd is designed to store a small amount of metadata in consistent, fault-tolerant way. Distributed systems use etcd as the root of scheduling and service discovery configuration storage. [Distributed lock][etcd-concurrency] is implemented on top of etcd, useful for master election in other distributed databases.
+
+# What does etcd stand for?
```
Comment (xiang90) [inline Documentation/learning/why.md]: remove this section. we probably should just mention it in the way part about its origin.

File: Documentation/learning/why.md [hunk 6]
```diff
@@ -0,0 +1,21 @@
...
+
+TODO(fanmin)
+
+# Who uses etcd: container linux by CoreOS
+
+Application running on [container linux][container-linux] gets automatic, no-downtime Linux kernel updates. CoreOS uses [locksmith] to coordinate updates. locksmith stores semaphore values in etcd to ensure that only subset of cluster are rebooting at any given time.
```
Comment (xiang90) [inline Documentation/learning/why.md]: zero-downtime.

locksmith uses etcd to implement distributed semaphore, ensuring that ...

File: Documentation/learning/why.md [hunk 7]
```diff
@@ -0,0 +1,21 @@
...
+
+# Who uses etcd: Kubernetes
+
+[Kubernetes][kubernetes] needs configuration storage for service discovery and cluster management. Kubernetes API server writes cluster states in etcd as persistent storage. And watch API to monitor, roll out critical configuration changes. Consistency is the key to ensure that services correctly schedule and operate.
```
Comment (xiang90) [inline Documentation/learning/why.md]: needs ... -> uses etcd for xxx.

**Comment by gyuho** [comment]
All fixed. PTAL. Thanks.

**Comment by xiang90** [comment]
defer to @fanminshi and @heyitsanthony

**Comment by fanminshi** [comment]
LGTM. I think we should brain storm what we want to put in the why.md sometime in the future. I think providing real world use cases using etcd will be educational also.

File: Documentation/learning/why.md [hunk 8]
```diff
@@ -0,0 +1,14 @@
+# Why etcd
+
+etcd is designed to store a small amount of metadata in consistent, fault-tolerant way. Distributed systems use etcd as the root of scheduling and service discovery configuration storage. [Distributed lock][etcd-concurrency] is implemented on top of etcd, useful for master election in other distributed databases.
+
+Example use cases:
```
Comment (heyitsanthony) [inline Documentation/learning/why.md]: `## Use cases`?

File: Documentation/learning/why.md [hunk 9]
```diff
@@ -0,0 +1,14 @@
...
+
+- container linux by CoreOS: Application running on [container linux][container-linux] gets automatic, zero-downtime Linux kernel updates. CoreOS uses [locksmith] to coordinate updates. locksmith uses etcd to implement distributed semaphore, ensuring that only subset of cluster are rebooting at any given time.
```
Comment (heyitsanthony) [inline Documentation/learning/why.md]: s/container linux/Container Linux
s/CoreOS uses/Container Linux uses

locksmith implements a distributed semaphore over etcd to ensure only a subset of a cluster is rebooting at any given time

File: Documentation/learning/why.md [hunk 10]
```diff
@@ -0,0 +1,14 @@
...
+- [Kubernetes][kubernetes] uses etcd for configuration storage for service discovery and cluster management. Kubernetes API server writes cluster states in etcd as persistent storage. And watch API to monitor, roll out critical configuration changes. Consistency is the key to ensure that services correctly schedule and operate.
```
Comment (heyitsanthony) [inline Documentation/learning/why.md]: [Kubernetes][kubernetes] stores configuration data into etcd for service discovery and cluster management;  etcd's consistency is crucial for correctly scheduling and operating services. The Kubernetes API server persists cluster state into etcd. It uses etcd's watch API to monitor the cluster and roll out critical configuration changes.

File: Documentation/learning/why.md [hunk 11]
```diff
@@ -0,0 +1,14 @@
+# Why etcd
+
+etcd is designed to store a small amount of metadata in consistent, fault-tolerant way. Distributed systems use etcd as the root of scheduling and service discovery configuration storage. [Distributed lock][etcd-concurrency] is implemented on top of etcd, useful for master election in other distributed databases.
```
Comment (heyitsanthony) [inline Documentation/learning/why.md]: etcd stores metadata in a consistent and fault-tolerant way. Distributed systems use etcd as a consistent key-value store for configuration management, service discovery, and coordinating distributed work. Common distributed patterns using etcd include leader election, [distributed locks][etcd-concurrency], and monitoring machine liveness.

**Comment by gyuho** [comment]
All fixed. PTAL. Thanks!

File: Documentation/docs.md [hunk 12]
```diff
@@ -42,11 +42,12 @@ Administrators who need to create reliable and scalable key-value stores for the
 
 To learn more about the concepts and internals behind etcd, read the following pages:
 
- - Why etcd (TODO)
+ - [Why etcd][why] (TODO)
  - [Understand data model][data_model]
  - [Understand APIs][understand_apis]
  - [Glossary][glossary]
  - Internals (TODO)
+ - Features and system comparisons (TODO)
```
Comment (heyitsanthony) [inline Documentation/docs.md]: this goes in why.md?

**Comment by heyitsanthony** [comment]
lgtm after moving the comparison todo into why.md

**Comment by gyuho** [comment]
Moved. Thanks!

**Pull request merged**